File: sprint8/JS/lesson5code2.js
function prefix_function(s) {
  const N = s.length;
  const π = new Array(N);
  π[0] = 0;

  for (let i = 1; i < N; i++) {
    let k = π[i - 1];
    while (k > 0 && s.charAt(k) !== s.charAt(i)) {
      k = π[k - 1];
    }
    if (s.charAt(k) === s.charAt(i)) {
      k++;
    }
    π[i] = k;
  }
  return π;
}

File: sprint8/C#/lesson5code2.cs
int[] prefix_function(string s)
{
    int N = s.Length;
    int[] π = new int[N];
    for (int i = 1; i < N; i++)
    {
        int k = π[i - 1];
        while (k > 0 && s[k] != s[i])
        {
            k = π[k - 1];
        }
        if (s[k] == s[i])
        {
            k++;
        }
        π[i] = k;
    }
    return π;
}